storage/src/persistent/commit_log/mod.rs

```diff
@@ -16,23 +15,185 @@ use crate::persistent::schema::{CommitLogDescriptor, CommitLogSchema};
+        let data = &self.buf[self.acc..(self.acc + index.uncompressed_data_length as usize)];
+        self.acc += index.uncompressed_data_length as usize;
+        self.cursor += 1;
+        Some(data.to_vec())
+    }
+}
 
 /// Possible errors for commit log
 #[derive(Debug, Fail)]
 pub enum CommitLogError {
     #[fail(display = "Schema error: {}", error)]
     SchemaError { error: SchemaError },
-    #[fail(display = "Failed to append record: {}", error)]
-    AppendError { error: AppendError },
-    #[fail(display = "Failed to read record at {}: {}", location, error)]
-    ReadError {
-        error: ReadError,
-        location: Location,
-    },
     #[fail(display = "Commit log I/O error {}", error)]
     IOError { error: io::Error },
     #[fail(display = "Commit log {} is missing", name)]
     MissingCommitLog { name: &'static str },
+    #[fail(display = "Failed to read record at {}", location)]
+    ReadError { location: Location },
+    #[fail(display = "Failed to read record data corrupted")]
+    CorruptData,
+    #[fail(display = "Tezedge CommitLog error: {:#?}", error)]
+    TezedgeCommitLogError { error: TezedgeCommitLogError },
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct Index {
+    pub position: u64,
+    pub compressed_data_length: u64,
+    pub uncompressed_data_length: u64,
+}
+
+impl Index {
+    fn from_buf(buf: &[u8]) -> Result<Self, TezedgeCommitLogError> {
+        if buf.len() != TH_LENGTH {
+            return Err(TezedgeCommitLogError::IndexLengthError);
+        }
+        let mut buf = buf.to_vec();
+        let position_raw_bytes: Vec<_> = buf.drain(..8).collect();
+        let compressed_data_length_raw_bytes: Vec<_> = buf.drain(..8).collect();
+        let uncompressed_data_length_raw_bytes: Vec<_> = buf.drain(..).collect();
+        let position = u64::from_be_bytes(
+            position_raw_bytes
+                .as_slice()
+                .try_into()
+                .map_err(|_| TezedgeCommitLogError::TryFromSliceError)?,
+        );
+        let compressed_data_length = u64::from_be_bytes(
+            compressed_data_length_raw_bytes
+                .as_slice()
+                .try_into()
+                .map_err(|_| TezedgeCommitLogError::TryFromSliceError)?,
+        );
+        let uncompressed_data_length = u64::from_be_bytes(
+            uncompressed_data_length_raw_bytes
+                .as_slice()
+                .try_into()
+                .map_err(|_| TezedgeCommitLogError::TryFromSliceError)?,
+        );
+        Ok(Self {
+            position,
+            compressed_data_length,
+            uncompressed_data_length,
+        })
+    }
+    fn new(position: u64, compressed_data_length: u64, uncompressed_data_length: u64) -> Self {
+        Self {
+            position,
+            compressed_data_length,
+            uncompressed_data_length,
+        }
+    }
+    fn to_vec(&self) -> Vec<u8> {
+        let mut buf = vec![];
+        for b in &self.position.to_be_bytes() {
```

Comment:
Shorter and probably faster:

```rust
 buf.extend_from_slice(&self.position.to_be_bytes());
```

Reply:
Yeah, completely forgot about that function

Reply:
Thanks